awscli/autocomplete/autogen.py

```diff
@@ -0,0 +1,177 @@
+            resource = self._resource_for_single_operation(
+                op_name, service_model)
+            if resource is not None:
+                self._inject_resource(all_resources, resource)
+        return all_resources
+
+    def _generate_operations(self, op_names, resources, service_model):
+        # member_name -> resource
+        reverse_mapping = {}
+        for name, identifier_map in resources.items():
+            for key in identifier_map['resourceIdentifier']:
+                reverse_mapping[key] = name
+        op_map = {}
+        for op_name in op_names:
+            op_model = service_model.operation_model(op_name)
+            input_shape = op_model.input_shape
+            if not input_shape:
+                continue
+            for member in input_shape.members:
+                member_name = self._find_matching_member_name(
+                    member, reverse_mapping)
+                if member_name is None:
+                    continue
+                resource_name = reverse_mapping[member_name]
+                op = op_map.setdefault(op_name, {})
+                param = op.setdefault(member, {})
+                param['completions'] = [
+                    {'parameters': {},
+                     'resourceName': resource_name,
+                     'resourceIdentifier': member}
+                ]
+        return op_map
+
+    def _find_matching_member_name(self, member, reverse_mapping):
+        # Try to find something in the reverse mapping that's close
+        # to the member name.  This is really inefficient but this is
+        # a build time step so we don't really care about perf.
+        if member in reverse_mapping:
+            return member
+        matcher = SequenceMatcher()
+        matcher.set_seq2(member)
+        for candidate in reverse_mapping:
+            matcher.set_seq1(candidate)
+            match_ratio = matcher.ratio()
+            if match_ratio > 0.9:
+                return candidate
+
+    def _inject_resource(self, all_resources, resource):
+        r = all_resources.setdefault(resource.resource_name, {})
+        r['operation'] = resource.operation
+        ident = r.setdefault('resourceIdentifier', {})
+        ident[resource.ident_name] = resource.jp_expr
+
+    def _resource_for_single_operation(self, op_name, service_model):
+        # This heuristic works if there's a List/Describe operation
+        # that has only one shape in the output that's of type list.
+        # It will then find that inspect that list member to find
+        # the best guess at which identifier maps to the resource.
+        # This is a best effort attempt and relies heavily on naming
+        # conventions.
+        op_model = service_model.operation_model(op_name)
+        output = op_model.output_shape
+        list_members = [member for member, shape in output.members.items()
+                        if shape.type_name == 'list']
+        if len(list_members) != 1:
```

Comment:
How again do we model resources that have a required parameter in order to list (i.e. subresource)?

Reply:
Via the `parameters` key.  Right now we always generate an empty map so we're not taking advantage of this feature.